pkg/asyncui/asyncui.go

```diff
@@ -0,0 +1,143 @@
+package asyncui
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/components"
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment"
+	"github.com/kyma-project/cli/pkg/step"
+)
+
+// StepFactory is a factory used to generate a step in the UI.
+type StepFactory interface {
+	NewStep(msg string) step.Step
+}
+
+// End-user messages
+const (
+	deployPrerequisitesPhaseMsg   string = "Deploying pre-requisites"
+	undeployPrerequisitesPhaseMsg string = "Undeploying pre-requisites"
+	deployComponentsPhaseMsg      string = "Deploying Kyma"
+	undeployComponentsPhaseMsg    string = "Undeploying Kyma"
+	deployComponentMsg            string = "Deploying component '%s'"
+)
+
+// AsyncUI renders the CLI ui based on receiving events
+type AsyncUI struct {
+	// used to create UI steps
+	StepFactory StepFactory
+	// channel provided by caller to retrieve errors
+	Errors chan<- error
```

Comment:
Do those 2 struct fields have to be exposed? I guess, they could be kept internal?

Reply:
Good point, thanks for bringing it up. The purpose of this field is to provide users of the Async-UI the option to retrieve errors / failures. Right now, the failures are just leading to failing steps but not communicated up to the caller of the Async-UI.

But it's worth to think whether this is:

- required at all
- needs further explanations (e.g. enhancing the comment or renaming the field name...)